# pandapower WLS State Estimation
This is an introduction into the usage of the pandapower state estimation module. It shows how to 
create measurements for a pandapower network and how to perform a state estimation with the weighted least squares (WLS) algorithm.

## Example Network

We will be using the reference network from the book "Power System State Estimation" by Ali Abur and Antonio Gómez Expósito. 
It contains 3 buses with connecting lines between buses 1-2, 1-3 and 2-3. 8 measurements of different types enable WLS state estimation.

We first create this network in pandapower.

In [1]:
import pandapower as pp

net = pp.create_empty_network()

b1 = pp.create_bus(net, name="bus 1", vn_kv=1., index=1)
b2 = pp.create_bus(net, name="bus 2", vn_kv=1., index=2)
b3 = pp.create_bus(net, name="bus 3", vn_kv=1., index=3)

pp.create_ext_grid(net, 1)  # set the slack bus to bus 1

l1 = pp.create_line_from_parameters(net, 1, 2, 1, r_ohm_per_km=.01, x_ohm_per_km=.03, c_nf_per_km=0., imax_ka=1)
l2 = pp.create_line_from_parameters(net, 1, 3, 1, r_ohm_per_km=.02, x_ohm_per_km=.05, c_nf_per_km=0., imax_ka=1)
l3 = pp.create_line_from_parameters(net, 2, 3, 1, r_ohm_per_km=.03, x_ohm_per_km=.08, c_nf_per_km=0., imax_ka=1)

net

This pandapower network includes the following parameter tables:
   - line (3 elements)
   - bus (3 elements)
   - ext_grid (1 elements)

## Adding Measurements

Measurements are defined via the pandapower *create_measurement* function.    
The physical properties which can be measured are set with the *type* argument and can be one of the following: "p" for active power, "q" for reactive power, "v" for voltage and "i" for electrical current.   
The element is set with the *element_type* argument, it can be either "bus", "line" or "transformer".    
Power is measured in kW / kVar, voltage in per unit and current in A. Bus power injections are positive if power is generated at the bus and negative if it is consumed.

Now we can add our measurements, which are valid for one point in time.

We add two voltage magnitude measurements on buses 1 / 2 with voltage magnitude of 1.006 pu / 0.968 pu and a standard deviation of 0.004 pu each:

In [2]:
pp.create_measurement(net, "v", "bus", 1.006, .004, b1)        # V at bus 1
pp.create_measurement(net, "v", "bus", 0.968, .004, b2)        # V at bus 2
net.measurement

type element_type  value  std_dev  bus element
0    v          bus  1.006    0.004    1    None
1    v          bus  0.968    0.004    2    None

We add bus injection measurements on bus 2 with P=-501 kW and Q=-286kVar and standard deviations of 10kVA: 

In [3]:
pp.create_measurement(net, "p", "bus", -501, 10, b2)         # P at bus 2
pp.create_measurement(net, "q", "bus", -286, 10, b2)         # Q at bus 2
net.measurement

type element_type    value  std_dev  bus element
0    v          bus    1.006    0.004    1    None
1    v          bus    0.968    0.004    2    None
2    p          bus -501.000   10.000    2    None
3    q          bus -286.000   10.000    2    None

Finally, we add line measurements for lines 0 and 1, both placed at the side of bus 1. The bus parameter defines the bus at which the line measurement is positioned, the line argument is the index of the line.

In [4]:
pp.create_measurement(net, "p", "line", 888, 8, bus=b1, element=l1)    # Pline (bus 1 -> bus 2) at bus 1
pp.create_measurement(net, "p", "line", 1173, 8, bus=b1, element=l2)   # Pline (bus 1 -> bus 3) at bus 1
pp.create_measurement(net, "q", "line", 568, 8, bus=b1, element=l1)    # Qline (bus 1 -> bus 2) at bus 1
pp.create_measurement(net, "q", "line", 663, 8, bus=b1, element=l2)    # Qline (bus 1 -> bus 3) at bus 1
net.measurement

type element_type     value  std_dev  bus element
0    v          bus     1.006    0.004    1    None
1    v          bus     0.968    0.004    2    None
2    p          bus  -501.000   10.000    2    None
3    q          bus  -286.000   10.000    2    None
4    p         line   888.000    8.000    1       0
5    p         line  1173.000    8.000    1       1
6    q         line   568.000    8.000    1       0
7    q         line   663.000    8.000    1       1

## Performing the State Estimation

The measurements are now set. We have to initialize the starting voltage magnitude and voltage angles for the state estimator. In continous operation, this can be the result of the last state estimation. In our case, we set flat start conditions: 1.0 p.u. for voltage magnitude, 0.0 degree for voltage angles. This is easily done with the parameter "init", which we define as "flat".

And now run the state estimation. Afterwards, the result will be stored in the table res_bus_est.

In [5]:
from pandapower.estimation import estimate

success = estimate(net, init='flat')
print(success)

True


## Working with Results

We can show the voltage magnitude and angles directly:

In [6]:
net.res_bus_est.vm_pu

1    0.999629
2    0.974156
3    0.943890
Name: vm_pu, dtype: float64

In [7]:
net.res_bus_est.va_degree

1    0.000000
2   -1.247547
3   -2.745717
Name: va_degree, dtype: float64

The results match exactly with the results from the book: Voltages 0.9996, 0.9742, 0.9439; Voltage angles 0.0, -1.2475, -2.7457). Nice!    
Let's look at the bus power injections, which are available in res_bus_est as well

In [8]:
net.res_bus_est.p_kw

1   -2064.016454
2     495.974966
3    1514.220983
Name: p_kw, dtype: float64

In [9]:
net.res_bus_est.q_kvar

1   -1226.440397
2     297.749528
3     787.528929
Name: q_kvar, dtype: float64

We can also compare the resulting line power flows with the measurements.

In [10]:
net.res_line_est.p_from_kw

0     892.991989
1    1171.024465
2     385.911626
Name: p_from_kw, dtype: float64

In [11]:
net.res_line_est.q_from_kvar

0    558.821689
1    667.618708
2    227.755971
Name: q_from_kvar, dtype: float64

Again, this values do match the estimated values from our reference book.
This concludes the small tutorial how to perform state estimation with a pandapower network.